File: host/src/models/cache.ts
/** The model cache: one directory of files, each identified by its sha256. */
import { statSync } from 'node:fs'
import { join } from 'node:path'
import { err, ok, type Result } from '../result.ts'
import { sha256Of } from './digest.ts'
import { integrity, invalidEntry, missing, unreadable, type ModelsError } from './errors.ts'
import { defaultRoot } from './root.ts'
import { modelEntryContract, type ResolvedModel } from './schema.ts'

export class Cache {
  #root: string

  constructor(root: string) {
    this.#root = root
  }

  /** Cache at `GAME_BOX_MODELS_DIR`, or the platform cache directory. */
  static open(): Cache {
    return new Cache(defaultRoot())
  }

  /** Cache at an explicit directory. */
  static at(root: string): Cache {
    return new Cache(root)
  }

  get root(): string {
    return this.#root
  }

  /** Locate a catalog entry and verify its digest. Nothing comes back unverified. */
  async resolve(entry: unknown): Promise<Result<ResolvedModel, ModelsError>> {
    const parsed = modelEntryContract.parse(entry)
    if (!parsed.ok) return err(invalidEntry(parsed.error))

    const expected = parsed.value.sha256
    const path = join(this.#root, parsed.value.file)
    if (!isFile(path)) return err(missing(path))

    let digest
    try {
      digest = await sha256Of(path)
    } catch (cause) {
      return err(unreadable(cause))
    }
    if (digest.sha256 !== expected) return err(integrity(expected, digest.sha256))

    return ok({ id: parsed.value.id, path, sizeBytes: digest.sizeBytes, sha256: digest.sha256 })
  }
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile()
  } catch {
    return false
  }
}

File: models/src/cache.rs
//! The model cache: one directory of files, each identified by its sha256.

use crate::digest::sha256_of;
use crate::error::ModelsError;
use crate::root::default_root;
use crate::schema;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};

pub struct Cache {
    root: PathBuf,
}

impl Cache {
    /// Cache at `GAME_BOX_MODELS_DIR`, or the platform cache directory.
    pub fn open() -> Self {
        Cache { root: default_root() }
    }

    /// Cache at an explicit directory.
    pub fn at(root: impl Into<PathBuf>) -> Self {
        Cache { root: root.into() }
    }

    pub fn root(&self) -> &Path {
        &self.root
    }

    /// Locate a catalog entry (`schema/model-entry.json`) and verify its digest.
    /// Returns a resolved model (`schema/resolved-model.json`).
    pub fn resolve(&self, entry: &Value) -> Result<Value, ModelsError> {
        schema::entry()
            .validate(entry)
            .map_err(|e| ModelsError::InvalidEntry(e.to_string()))?;
        let expected = entry["sha256"].as_str().unwrap_or_default().to_string();
        let path = self.root.join(entry["file"].as_str().unwrap_or_default());
        if !path.is_file() {
            return Err(ModelsError::Missing(path));
        }
        let (actual, size) = sha256_of(&path).map_err(|e| ModelsError::Unreadable(e.to_string()))?;
        if actual != expected {
            return Err(ModelsError::Integrity { expected, actual });
        }
        let resolved = json!({
            "id": entry["id"],
            "path": path.to_string_lossy(),
            "sizeBytes": size,
            "sha256": actual
        });
        debug_assert!(schema::resolved().is_valid(&resolved), "resolved off-contract: {resolved}");
        Ok(resolved)
    }
}
